pjrt-plugin/iree/integrations/pjrt/common/api_impl.cc

```diff
@@ -438,19 +457,83 @@ iree_status_t BufferInstance::GetHostSizeInBytes(iree_host_size_t* host_size) {
   return iree_ok_status();
 }
 
+iree_status_t BufferInstance::AsyncDeallocate() {
+  IREE_TRACE_SCOPE();
+  return iree_hal_device_queue_dealloca(
+      device().device(), IREE_HAL_QUEUE_AFFINITY_ANY,
+      /*wait_semaphore_list=*/iree_hal_fence_semaphore_list(done_fence()),
+      /*signal_semaphore_list=*/iree_hal_semaphore_list_empty(),
+      iree_hal_buffer_view_buffer(buffer_view_.get()));
+  return iree_ok_status();
+}
+
 iree_status_t BufferInstance::CopyToHost(void* dst, iree_host_size_t dst_size,
-                                         EventInstance** done_event) {
-  // TODO: Do an async transfer on a transfer queue like a grown up.
-  iree_hal_device_t* hal_device;
-  IREE_RETURN_IF_ERROR(device_.GetHalDevice(&hal_device));
-  IREE_RETURN_IF_ERROR(iree_hal_device_transfer_d2h(
-      hal_device, iree_hal_buffer_view_buffer(buffer_view()), 0, dst, dst_size,
-      IREE_HAL_TRANSFER_BUFFER_FLAG_DEFAULT, iree_infinite_timeout()));
-
-  *done_event = new EventInstance();
+                                         EventInstance** out_done_event) {
+  // Set up an event for external triggering. While a little wonky, we
+  // trigger it in the host buffer release callback, which happens once the
+  // transfer is done. I don't love this option but it seems to match what
+  // I'm looking for.
+  EventInstance* capture_done_event =
+      new EventInstance(EventInstance::Type::EXTERNAL);
+  *out_done_event = capture_done_event;
+
+  // Import the destination (host) buffer as an iree_hal_buffer_t so that we
+  // can issue copy commands.
+  iree::vm::ref<iree_hal_buffer_t> dst_buffer;
+  iree_hal_buffer_params_t dst_buffer_params = {
```

Comment:
you'll want to set the type to IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE